rmw_fastrtps_cpp/src/rmw_client.cpp

```diff
@@ -124,25 +122,23 @@ rmw_create_client(
   subscriberParam.topic.topicKind = NO_KEY;
   subscriberParam.topic.topicDataType = response_type_name;
   subscriberParam.historyMemoryPolicy = PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
-  rcutils_ret_t ret = _assign_partitions_to_attributes(
-    service_name, ros_service_response_prefix,
-    qos_policies->avoid_ros_namespace_conventions, &subscriberParam);
-  if (ret != RCUTILS_RET_OK) {
-    // error msg already set
-    goto fail;
+  if (!qos_policies->avoid_ros_namespace_conventions){
+  subscriberParam.topic.topicName = std::string(ros_service_response_prefix) + service_name;
```

Comment:
I think there should be a "/" used to separate the ros prefix and the service/topic name (at least that's what the code in https://github.com/ros2/rmw_fastrtps/pull/192/files#diff-0ce60ac271e069c6686c50d38559e0a5L42 was doing for the string.
e.g. ROS topic `foo` => DDS topic `rt/foo`

Reply:
The code in https://github.com/ros2/rmw_fastrtps/pull/192/files#diff-0ce60ac271e069c6686c50d38559e0a5L42 is first extracting the base topic_name from the whole topic name (which includes the namespaces), by taking whatever is there after the last '/'. And then re-adding prefix to the base topic_name along with  '/'. So it will have ros prefix + '/' + base name. But I guess here the default(complete) topic name which we receive will have a "/", isn't it?  example: we will have `/chatter` by default rather than  `chatter`, otherwise I'll have add a check to check if there is a `/` by default or not and then add it to the topic name.

Reply:
> But I guess here the default(complete) topic name which we receive will have a "/", isn't it?

Apparently if called from rcl, it always has a leading forward slash as [rcl enforces it](https://github.com/ros2/rcl/blob/d41c923927850274f63aee1bde565d3a8266a7bd/rcl/src/rcl/publisher.c#L148) before passing the name to rmw.

Not sure if we can make that assumption for people calling rmw API directly though.
So my guess (maybe someone from @ros2/team more familiar with the matter can pitch in) is that, if `avoid_ros_namespace_conventions` is false, we should check for the presence of the slash and add it if needed.
If `avoid_ros_namespace_conventions` is true, we should not modify the name and let user creat topics with the nam he specified.

Reply:
By the time it is passed into `rmw` all topics will be fully qualified, which includes a leading `/`.

Reply:
> By the time it is passed into rmw all topics will be fully qualified

I agree with that, I was just wondering if it's safe to make this assumption here for people calling this API directly (and not via rcl).
I guess the goal here is to avoid adding the namespace prefix but still assume ROS naming conventions are respected. Does that mean that there is no way to subscribe to a DDS topic named e.g. "image" using this API ? only "/image" ?

Reply:
> I was just wondering if it's safe to make this assumption here for people calling this API directly (and not via rcl).

No one should do that 😄, but if I were to write the documentation for these rmw functions now I'd say it is expected that the topic name is good as-is and needs no modification. So basically, prefix with `rt__` or w/e if `avoid_ros_namespace_conventions` is true, otherwise do nothing to the topic name.

Separately the use case of subscribing to a DDS topic `foo` rather than `/foo`, should be handled. I hadn't thought about that, because in the previous case you simply used the base name you wanted and that was it. There were ideas in the design document about how to handle that, but nothing was implemented:

http://design.ros2.org/articles/topic_and_service_names.html#examples-and-ideas-for-communicating-with-non-ros-topics

Even if you were to strip leading `/` when `avoid_ros_namespace_conventions` is true, you still would not be able to specify a partition name.

Reply:
>  So basically, prefix with rt__ or w/e if avoid_ros_namespace_conventions is true, otherwise do nothing to the topic name.

Sounds good to me

> Even if you were to strip leading / when avoid_ros_namespace_conventions is true, you still would not be able to specify a partition name.

Fair enough.

Ideally people could use whatever is exposed in the API, in this case we don't expose partitions to users, so users would only be able to be matched with a topic that has matching QoS (so only if empty partitions on both sides).
We can definitely postpone the discussion on how to enable users to subscribe to pure DDS topics :+1:

Reply:
So are we deciding on leaving it as is? This will affect the topic names coming in the future which don't start with a "/". Or some remapping might be required to append a "/' for those topic or service names.

Reply:
As we don't have any currently supported use case where the topic will not enter this function without a leading slash we keep it as is for now :+1: . If the use case comes up we'll adapt the behavior.

Reply:
I was thinking about this more, and I suppose we could choose to pass the unexpanded topic name directly to the rmw layer if the `avoid_ros_namespace_conventions` is set to true.

For example, we currently:

- pass topic name into `rcl_create_publisher()` (for example), e.g. `image`
- expand topic name with `rcl_expand_topic_name()`, e.g. `image` -> `/image`
- store expanded topic name in `publisher.topic_name`
- pass `publisher.topic_name` to `rmw_create_publisher()`

And instead, iff `avoid_ros_namespace_conventions` is true we could:

- pass topic name into `rcl_create_publisher()` (for example), e.g. `image`
- ~~expand topic name with `rcl_expand_topic_name()`, e.g. `image` -> `/image`~~
- store ~~expanded~~ topic name in `publisher.topic_name`
- pass `publisher.topic_name` to `rmw_create_publisher()`
  - `rmw_create_publisher()` would also not prefix with the ROS prefix

The only thing I'd be worried about is what does `rcl_publisher_get_topic_name()` return? The expanded topic name or the one used in `rmw_create_publisher()`?

----

I don't mean to delay this pull request, but I think we need to resolve this and update the design doc before committing to changes in the source.

Reply:
Agreed.
I suggest letting this set of PRs move forward and have first an update to the design doc to remove the partitions and describe the new behavior from from this set of PRs in the "ROS case" ([reported on the parent issue](https://github.com/ros2/ros2/issues/476#issuecomment-378408720).
We can have a separate discussion regarding what should happen in the "avoid_ros_namespace_conventions" case, update the design doc and propose code changes once design doc approved.